### PR TITLE
records.config to records.yaml

### PR DESCRIPTION
**Pull request by brbzull0**
### Description

This PR introduces  the change from a records style configuration structure to a YAML structure(records.yaml).

There are a few commits in this PR which are separated so we can split this into smaller PRs needed, also to be able to read the change in small chunks.

A few things that needs to be highlighted 

### New YAML structure

We are introducing a root node `ts` for the `records.yaml`, so all existing records will belong to the `ts` node.
From the current `records.config` structure we  have dropped  the prefix `proxy.config` and `proxy.local` and use the rest to build a YAML structure.
The logic is basically to walk down the record name separated by the dots and build a new YAML map from each name(separated by dots),  so for instance, the following records:

```
CONFIG proxy.config.exec_thread.autoconfig.scale FLOAT 1.0
CONFIG proxy.config.exec_thread.limit INT 2
CONFIG proxy.config.accept_threads INT 1
CONFIG proxy.config.task_threads INT 2
CONFIG proxy.config.cache.threads_per_disk INT 8
CONFIG proxy.config.exec_thread.affinity INT 1
CONFIG proxy.config.diags.debug.enabled INT 0
CONFIG proxy.config.diags.debug.tags STRING http|dns
```
We first drop the `proxy.config.` and walk down each field, so we will end up with the following:

```yaml
ts:
  accept_threads: 1
  cache:
    threads_per_disk: 8
  diags:
    debug:
      enabled: 0
      tags: http|dns
  exec_thread:
    affinity: 1
    autoconfig:
      scale: 1.0
    limit: 2
  task_threads: 2
```
There were a few things that needed to be done on the names and types that I will be describing next.

### Records renaming

There are few changes that I had to perform in order to be able to keep the similar structure used in the `records.config` into a YAML based configuration.

An example of this would be a record key holding a value which also contains children fields:

```
proxy.config.exec_thread.autoconfig INT 1
proxy.config.exec_thread.autoconfig.scale FLOAT 1.0 << children of autoconfig
```
As this cannot be done in YAML we need to rename `proxy.config.exec_thread.autoconfig` to `proxy.config.exec_thread.autoconfig.enabled` so we no longer have a map with values other than some children fields. So in YAML we can go from something like this:

```
proxy.config.exec_thread.autoconfig INT 1
proxy.config.exec_thread.autoconfig.scale FLOAT 1.0
```
to 
```yaml
exec_thread:
    autoconfig:
      enabled: 1 << new field.
      scale: 1.0
```

All this record renaming  will be reflected in the ATS core as well.

> **_NOTE:_**   The convert to yaml script will work away all this.

#### Naming convention

All this renamed records are subject to the following reasoning:

If a value is meant to be used as a toggle on/off, I’ve used the name`enabled`.
If a value is meant to be used a an enumeration, I’ve used the name `value`

Possible issue with this is that if we scale the field meaning and we need to move from a toggle to an enumeration then this logic will make no sense. 
Should we just use one or the other? Current `diags.debug.enabled` is an example of an enable field holding more than `0,1` value.

#### records remaning
The following is the list records that needs to be changed:

```
proxy.config.output.logfile -> proxy.config.output.logfile.name
proxy.config.exec_thread.autoconfig -> proxy.config.exec_thread.autoconfig.enabled
proxy.config.hostdb -> proxy.config.hostdb.enabled
proxy.config.tunnel.prewarm -> proxy.config.tunnel.prewarm.enabled
proxy.config.ssl.origin_session_cache -> proxy.config.ssl.origin_session_cache.enabled
proxy.config.ssl.session_cache -> proxy.config.ssl.session_cache.value
proxy.config.ssl.TLSv1_3 -> proxy.config.ssl.TLSv1_3.enabled
proxy.config.ssl.client.TLSv1_3 -> proxy.config.ssl.client.TLSv1_3.enabled
```

### Record types

Currently the way for ATS to know the record type is by specifying it in the config file (`proxy.config.exec_thread.autoconfig INT 1`) and also inside  `RecordsConfig.cc` (not only the type but also name, type, default value, validation regex, etc). This internally lets us validate the config record set in the `records.config` among the one registered internally; this is done to avoid mismatched types, etc.

In YAML you do not need to explicitly set the  type as some of the types can be deduced, using the `YAML-CPP` library we are not able to deduce this type unless we use [tags](https://yaml.org/spec/1.2.2/#24-tags) for a field with the corresponding type (!!int, !!float, etc). 

> **_NOTE:_**  Is important to note that  for core registered records we do not need to set the type as we have them already in the  `RecordsConfig.cc` array.

#### Non core records.

Now the problem arises when a plugin register a new field which is not registered in the core, the legacy implementation would read the record type from the config file and then once the record is registered by a plugin it will be properly registered inside ATS( * mismatched type gets the current value reset to the default one*)

As said before in ATS when using YAML without explicit saying the type we have no way to know the type till the registration happens(after reading the config). We need to overcome this and to me there seems to be two options here:

1. Expect non core records to set the type (!!int, !!float, etc). 
```yaml
ts:
   plugin_x:
      my_field_1: !!int '1'
      my_field_2: !!float '1.2'
      my_field_3: 'my string'
```
2. Non core records  will be all treated as strings and once registered by a plugin we set the type right(this is already done) but we do **NOT** reset the value to default(as done currently), again this is a change that may cause issues if anyone is expecting the value to be reset.

This PR implements option 1.

> **_NOTE:_**  The convert tool provided to migrate from `records.config` to `records.yaml` has an option to include types in the converted file. All existing records.config files that will be migrated with this tool won't need to set the type, but new fields added to the already migrated file would.

**Just to be 100% clear, there is no need to specify the type on all the fields, just the ones that aren’t present in the core.**

##### Why are there no tags in any of the current yaml config files handled by ATS?

ATS is parsing all the YAML files by hand, this means that each field is known by  the decoder so it deals with the type properly. For records as we keep the internal librecords structure we do not deal with any particular field as this is already done by the librecords.

> **_NOTE:_** Further down the line a schema file should be introduced and replace the `RecordsConfig.cc` ; this may not help with non-core records, unless it’s changed by the user.

### Code changes

There is no change in the current librecords implementation, I have just added a bunch of functions to read the yaml (instead of parsing the legacy text file) and then  register the records(as currently done). Some of these functions are the base for a future TS API that can be used by `conf_remap`.(proposal in progress) and also they will be used by the JSON-RPC to be able to inject `records.yaml`.

As said, we keep the internal record names without a change, so querying and setting records keeps the same. 

How do we keep the record name from the YAML structure?

As described above this just just build down the YAML structure base on the record name, so to build up the record name from a YAML field we just need to walk it down and build the name, so the `records.yaml` parser will flatten the following:

```yaml
ts:
  exec_thread:
    autoconfig:
      scale: 1.0
```

to `proxy.config.exec_thread.autoconfig.scale` for instance. So this code is just a thin layer on top of the existing librecords API which still deals with records names as before.

Few notes:

- Traffic_ctl still works with record names, all existing API remains unchanged. Further down the line I will be adding a new JSON-RPC API to push YAML config. The existing code will work fine with this.
- All existing internal TS API remains unchanged.

**Important**

This implementation will only accept a `records.yaml` config file, if not found the defaults will be used. If a `record.config` file is found in the config directory a warning will be shown. This is done to avoid someone having the legacy file thinking it is the current one.

```
[Dec 20 18:42:12.075] traffic_server WARNING: Found a legacy config file. /home/to/ats/config/records.config
```

#### Unit test

Beside the records that were renamed I had changed the `trafficserver.ext` to keep the same interface for records update:

```python
       ts.Disk.records_config.update({
            'proxy.config.diags.logfile.filename': 'stdout',
            'proxy.config.error.logfile.filename': 'stdout',
        })
```
The above remains the same.

It usually takes a dict with the name and the value, this internally will be parsed and converted to a YAML file and then injected into the `record.yaml` file for the particular unit test.

There is also the possibility to just use yaml instead of a dict and update just the nodes passed in the YAML string.

```python
ts.Disk.records_config.update(
    '''
    diags:
      debug:
        enabled: 1
        tags: rpc|rec
    '''
 )
```

For the unit tests all this is transparent  and they can either keep using the records style of just the YAML, I would recommend for new tests just use YAML straight away.

I have introduced a new  file type `ats:config:yaml` which can be used to build any YAML style file.

### Documentation

This area is still WIP, but I am introducing a new `records.yaml` [RST](https://ci.trafficserver.apache.org/job/Github_Builds/job/docs/597/artifact/output/9264/docbuild/html/admin-guide/files/records.yaml.en.html) (in replace of the old one) which keeps the record name in the documentation but also includes the YAML representation of each record.

[ ] I still need to complete a guide to migrate from one to another.

### conf_remap plugin.

There will be a proposal API for plugins like conf_remap that needs to parse YAML files and work away from the record name as we do in core, this will be added after.

### CODE REVIEW

**Review by maskit — COMMENTED** [review]
I feel like we should probably gather feedback for setting names and the detailed format (I'm fine with YAML but see my inline comment) from users (or people who don't review this PR).

File: tools/records/convert2yaml.py [hunk 1]
```diff
@@ -0,0 +1,306 @@
...
+Remapped_Vars = {}
+
+Renamed_Records = {
+    # as it can have a numeric value [0-2]
+    'proxy.config.output.logfile': 'proxy.config.output.logfile.name',
+    'proxy.config.exec_thread.autoconfig': 'proxy.config.exec_thread.autoconfig.enabled',
+    'proxy.config.hostdb': 'proxy.config.hostdb.enabled',
+    'proxy.config.tunnel.prewarm': 'proxy.config.tunnel.prewarm.enabled',
+    'proxy.config.ssl.origin_session_cache': 'proxy.config.ssl.origin_session_cache.enabled',
+    'proxy.config.ssl.session_cache': 'proxy.config.ssl.session_cache.value',
```
[the remark this answers is not in the document]
Comment (brbzull0) [reply tools/records/convert2yaml.py:47]: > I feel like we should probably gather feedback for setting names and the detailed format (I'm fine with YAML but see my inline comment) from users (or people who don't review this PR).

Hello!!

yes definitelly, that's the plan.
I'll clarify all this once I complete the description with the reasoning behind the types/names,etc  and of course the idea is to gather feedback, I'll be also sending an email around with this renaming as I some may consider them breaking changes.

**Comment by brbzull0** [comment]
> I feel like we should probably gather feedback for setting names and the detailed format (I'm fine with YAML but see my inline comment) from users (or people who don't review this PR).

Hello!!

yes definitelly, that's the plan.
I'll clarify all this once I complete the description with the reasoning behind the types/names,etc and of course the idea is to gather feedback, I'll be also sending an email around with this renaming as I some may consider them breaking changes.

File: configs/records.yaml.default.in [hunk 2]
```diff
@@ -0,0 +1,218 @@
+##############################################################################
+# *NOTE*: All options covered in this file should be documented in the docs:
+#
+#    https://docs.trafficserver.apache.org/records.yaml
+##############################################################################
+
+ts:
+  accept_threads: !!int '1'
```
[the remark this answers is not in the document]
Comment (brbzull0) [reply configs/records.yaml.default.in:8]: We don't need this, there is only 1 scenario where this would be needed and this is when you register configuration inside a plugin which core has no idea about the type at loading time(before the plugin gets to say `RecRegisterConfigXX`). I will add a description on this PR about this in the next hours.

File: tools/records/convert2yaml.py (hunk 1 above)
[the remark this answers is not in the document]
Comment (brbzull0) [reply tools/records/convert2yaml.py:47]: Added some details on the PR description. Thanks.

**Comment by brbzull0** [comment]
[approve ci autest]

**Comment by brbzull0** [comment]
[approve ci]

**Comment by brbzull0** [comment]
[approve ci fedora]

**Comment by bneradt** [comment]
[approve ci fedora]

**Comment by brbzull0** [comment]
[approve ci debian]

**Comment by brbzull0** [comment]
[approve ci debian]

**Review by maskit — DISMISSED** [review]
I feel like we should merge this sooner rather than later, if we all agree to proceed this way at a minimum. Otherwise we may introduce new setting names that don't work with this, and then it requires updating this PR.

We can make setting name adjustments later. We may have some hiccups but we can find and fix them during 10.0 development. I'm approving this because I don't see any critical issue.

File: configs/records.yaml.default.in [hunk 3]
```diff
@@ -0,0 +1,218 @@
...
+    min_average_object_size: 8000
+
+##############################################################################
+# RAM and disk cache configurations. Docs:
+#    https://docs.trafficserver.apache.org/records.yaml#ram-cache
+#    https://docs.trafficserver.apache.org/en/latest/admin-guide/files/storage.config.en.html
+##############################################################################
+    ram_cache:
+      size: -1
+    ram_cache_cutoff: 4194304
```
Comment (SolidWallOfCode) [inline configs/records.yaml.default.in:32]:
```
ram_cache:
  size: -1
  cutoff: 4194304
```
Or maybe I should do separate PRs for these minor changes.

File: doc/ext/traffic-server.py [hunk 4]
```diff
@@ -67,12 +79,76 @@ class TSConfVar(std.Target):
...
+
+                    return False
+
+                if value is None or value == 'nullptr' or value == 'NULL':
+                    return None
+
+                # We want to make the type right when inserting the element into the dict.
+                if type == 'FLOAT':
+                    return float(value)
+                elif type == 'INT':
```
Comment (SolidWallOfCode) [inline doc/ext/traffic-server.py:106]: Might it be better to do a conversion and see if that works? Or convert to int and back and see if there's a difference?

File: mgmt/RecordsConfig.cc [hunk 5]
```diff
@@ -1530,3 +1530,15 @@ RecordsConfigIterate(RecordElementCallback callback, void *data)
     callback(&RecordsConfig[i], data);
   }
 }
+
+const RecordElement *
+GetRecordElementByName(std::string_view name)
```
Comment (SolidWallOfCode) [inline mgmt/RecordsConfig.cc:1535]: Where is this called? Is there a possible performance issue?

File: mgmt/config/FileManager.cc [hunk 6]
```diff
@@ -53,11 +53,11 @@ handle_file_reload(std::string const &fileName, std::string const &configName)
   ts::Errata ret;
   // TODO: make sure records holds the name after change, if not we should change it.
   if (fileName == ts::filename::RECORDS) {
-    if (RecReadConfigFile() == REC_ERR_OKAY) {
+    if (auto zret = RecReadYamlConfigFile(); zret) {
```
Comment (SolidWallOfCode) [inline mgmt/config/FileManager.cc:56]: Why is `zret` captured, if it's not used?

File: src/records/RecYAMLDecoder.cc [hunk 7]
```diff
@@ -0,0 +1,213 @@
...
+
+} // namespace
+
+namespace detail
+{
+void flatten_node(CfgNode const &field, RecYAMLNodeHandler handler, swoc::Errata &errata);
+std::pair<RecDataT, std::string> try_deduce_type(YAML::Node const &node);
+
+// Helper class to make the code less verbose when lock is needed.
+struct scoped_cond_lock {
```
Comment (SolidWallOfCode) [inline src/records/RecYAMLDecoder.cc:58]: Gah. Better to switch to `std::mutex` with `std::shared_lock` and `std::unique_lock`.

File: src/records/RecYAMLDecoder.cc [hunk 8]
```diff
@@ -0,0 +1,213 @@
...
+  RecSetRecord(rec_type, record_name.c_str(), data_type, &data, nullptr, source, false);
+  RecDataZero(data_type, &data);
+};
+
+swoc::Errata
+ParseRecordsFromYAML(YAML::Node root, RecYAMLNodeHandler handler, bool lock /*false by default*/)
+{
+  [[maybe_unused]] detail::scoped_cond_lock cond_lock(lock);
+
+  swoc::Errata errata;
```
Comment (SolidWallOfCode) [inline src/records/RecYAMLDecoder.cc:126]: You can dump this and use `return {};` as the final return statement.

File: src/records/RecYAMLDecoder.cc [hunk 9]
```diff
@@ -0,0 +1,213 @@
...
+    return swoc::Errata(ERRATA_NOTE, "'ts' root key not present or no fields to read.");
+  }
+
+  return errata;
+}
+
+namespace detail
+{
+std::pair<RecDataT, std::string>
+try_deduce_type(YAML::Node const &node)
```
Comment (SolidWallOfCode) [inline src/records/RecYAMLDecoder.cc:145]: I wonder if a `Lexicon` would work here.

**Review by SolidWallOfCode — APPROVED** [review]
Better to land this and fix the minor issues later.

**Comment by brbzull0** [comment]
[approve ci]

**Comment by brbzull0** [comment]
[approve ci autest]

**Comment by brbzull0** [comment]
[approve ci autest]

**Comment by brbzull0** [comment]
[approve ci autest]

**Comment by brbzull0** [comment]
[approve ci autest]

**Comment by brbzull0** [comment]
[approve ci autest]

**Comment by brbzull0** [comment]
[approve ci autest]
